addons/sale_project/models/sale_order_line.py

```diff
@@ -271,6 +271,17 @@ def _determine_project(so_line):
 
             return False
 
+        def _generate_milestone(sol):
+            if sol.product_id.service_policy in ['delivered_milestones']:
+                milestone = self.env['project.milestone'].create({
+                    'name': sol.name,
+                    'project_id': sol.project_id.id,
+                    'sale_line_id': sol.id,
+                    'quantity_percentage': 1,
+                })
+                if sol.product_id.service_tracking == 'task_in_project':
+                    sol.task_id.milestone_id = milestone.id
+
```

Comment:
Could you rather add this method in the model class? I know that it is like this above, but this ain't a good think in my opinion.